### ISSUE TITLE
Copying the envelope sender from the From header causes failures

### ISSUE THREAD

**Issue by robertmathews** [PHPMailer/PHPMailer#2298]
A recent change in commit 09bc3065290431cad8c345b852386789ca852396 makes PHPMailer pass the `-f` flag to sendmail in cases where it didn't use `-f` (i.e., where it previously let the system choose the envelope sender address / return path). This leads to failures in situations where it previously worked because the sendmail default envelope sender was used.

There's discussion of this in [this WordPress bug report](https://core.trac.wordpress.org/ticket/52822), but the overview is that (for example) many WordPress plugins send mail with a "From" header address that they don't really own.

This is obviously not best practice because it means DKIM won't work, but as long as `-f` is *not* passed to sendmail, the envelope sender address is likely to be valid. Either the system's default sendmail envelope sender address or the php.ini sendmail_from address will be used, and assuming one of those is correctly configured (which is usually the case with competent hosting companies, for example), the resulting mail will pass SPF checks and hopefully be delivered.

With this change, PHPMailer and WordPress send that message with `-f same_address_as_from_header`, overriding sendmail's default envelope sender. That means that SPF won't match for the message, either, and it's much more likely to be rejected as a forgery. (I've seen this problem happening in the wild after people updated to the version of WordPress that includes this PHPMailer change. Messages that used to be delivered are now rejected outright.)

In case an example helps, imagine a message was sent with "From: example@gmail.com" by a script that is running as user "user123" at "hostingcompany.com". Since the WordPress wp_mail() function calls PHPMailer without specifying a "sender", that message would previously be sent as:

```
Return-Path: user123@hostingcompany.com
From: example@gmail.com
```

This would probably work: While the "sending with a From header address you don't own" is not recommended for DKIM/DMARC purposes, at least the Return-Path envelope address is a valid address that can be matched using SPF, and it would likely be delivered. Now, however, the same message would be sent as:

```
Return-Path: example@gmail.com
From: example@gmail.com
```

Neither SPF or DKIM will be valid, so this message is much more likely to be rejected as a forgery.

It doesn't seem that the new "generating a default -f envelope sender from the header From address" behavior when no explicit sender is provided is a good idea. It may fix problems where the default PHP/sendmail binary envelope sender is misconfigured, but those have failed all along... and this change introduces new failures where the sendmail envelope default was properly configured.

If the envelope sender is unspecified when calling PHPMailer, it seems better to let PHP and the sendmail binary choose the default.

**Comment by Synchro** [comment]
Ugh. It's always the legacy stuff! This change was made primarily because it makes it act the same way as SMTP sending has done for a long time, and SMTP to localhost is always preferable to `mail()` because it's both faster and safer (`mail()` is inherently broken in that respect), as well as being more flexible and easier to debug. I also had bug reports going in exactly the opposite direction, that the default from address was causing non-delivery due to forgery, in a way that sending via SMTP didn't cause. Unfortunately a default address of `www-data@localhost.localdomain` is both common and useless.

The new code *does* use whatever is in `sendmail_from` but only if the `From` address is unset, which, to be fair, seems unlikely:

```php
        if ('' === $this->Sender) {
             $this->Sender = $this->From;
         }
         if (empty($this->Sender) && !empty(ini_get('sendmail_from'))) {
```

Would reversing the order of this check be better? So that the ini file setting takes precedence over the from address? Or do you think the only option (for WP) is to remove this mechanism altogether?

I suspect that this combo:

    Return-Path: user123@hostingcompany.com
    From: example@gmail.com

should result in rejection because gmail.com uses a hard `p=reject` DMARC, and there is no alignment here. I see a lot of people running into this on GoDaddy, because they block outbound SMTP and messages are routed through GD's servers, which will never pass Google's SPF. I think (don't know for sure) it's got bad enough that GD's mail server (`secureserver.net`) will actively reject messages if you try to send from domains like that.

**Comment by robertmathews** [comment]
> Ugh. It's always the legacy stuff!

Yup. Sorry.  :-(


>Unfortunately a default address of `www-data@localhost.localdomain` is both common and useless.

I agree with that. But many places have it set correctly, too, making it work for people who weren't previously complaining. I'd guess that people who run their own VPS, etc., often have it set incorrectly, but I think most large hosting companies (I work for a hosting company) do have it set to work, just because people complain like mad otherwise.

> Would reversing the order of this check be better? So that the ini file setting takes precedence over the from address? Or do you think the only option (for WP) is to remove this mechanism altogether?

Reversing the order of the check would fix it if there's an explicit php.ini setting, but I don't think that's common, and it wouldn't help in cases where it relies on the sendmail default working, which is probably more common.

> I suspect that this combo:
> 
> ```
> Return-Path: user123@hostingcompany.com
> From: example@gmail.com
> ```
> 
> should result in rejection because gmail.com uses a hard `p=reject` DMARC

Surprisingly, Gmail still actually uses `p=none`:

```
$ dig +short _dmarc.gmail.com TXT
"v=DMARC1; p=none; sp=quarantine; rua=mailto:mailauth-reports@google.com"
```

Other than the reports, that's supposed to be treated as if there were no DMARC policy present, so a matching SPF (even when not aligned in the DMARC sense) could have resulted in successful delivery before.

I'm certainly not arguing that it's reasonable for people to send from `@gmail.com` at all, ever. People should not do this, and if they used a reasonable From header address at a domain name they control, none of this would matter. So as you say, this is just a legacy question of what should happen for people who are doing it wrong. It's a question of whether it should work as it always did for people who have things set up badly.

>I see a lot of people running into this on GoDaddy, because they block outbound SMTP and messages are routed through GD's servers, which will never pass Google's SPF. I think (don't know for sure) it's got bad enough that GD's mail server (`secureserver.net`) will actively reject messages if you try to send from domains like that.

Right -- I would guess this change has made that worse, though, because the check they do to see if they'll allow it is probably against the envelope sender address, rather than the From header address. So I suspect they would have allowed this:

```
Return-Path: user123@server456.secureserver.net
From: example@gmail.com
```

But not:

```
Return-Path: example@gmail.com
From: example@gmail.com
```

(That is just a guess about GoDaddy, but I've seen that behavior from other companies.) Neither of those is great, but the first one has more of a chance of passing the outbound restrictions the company has, as well as possibly matching SPF.

**Comment by Synchro** [comment]
Dang, I could have sworn that gmail was stricter. Did a quick survey: Yahoo is reject, but Microsoft really can't decide what to do:

```
dig +short _dmarc.outlook.com TXT
"v=DMARC1; p=none; sp=quarantine; pct=100; rua=mailto:d@rua.agari.com; ruf=mailto:d@ruf.agari.com; fo=1"
dig +short _dmarc.office365.com TXT
"v=DMARC1; p=reject; pct=100; rua=mailto:d@rua.agari.com; ruf=mailto:d@ruf.agari.com; fo=1"
dig +short _dmarc.hotmail.com TXT
"v=DMARC1; p=none; rua=mailto:d@rua.agari.com;ruf=mailto:d@ruf.agari.com;fo=1:s:d"
```

>I'm certainly not arguing that it's reasonable for people to send from @gmail.com at all, ever. People should not do this, and if they used a reasonable From header address at a domain name they control, none of this would matter. So as you say, this is just a legacy question of what should happen for people who are doing it wrong. It's a question of whether it should work as it always did for people who have things set up badly.

Held to ransom by the past, again.

Do you think I should forget about handling `sendmail_from` at all, on the basis that if an envelope sender is not specified at all, PHP will use it anyway?

Would you like to make a PR to suggest code changes?

**Comment by galbaras** [comment]
> I also had bug reports going in exactly the opposite direction, that the default from address was causing non-delivery due to forgery

Admittedly, some of the discussion above got my eyes to glaze over, but a simple solution may be to provide a setting, which WordPress can set or unset, and which reverts the order of processing to the previous one, i.e. INI before From.

Having said that, since this was a change of behavior, seems to me that the default should go back to the previous behavior, and the new setting should activate the new one. Either way, a setting can help people decide what they want to do, and everybody can be happy 😄

**Comment by Synchro** [comment]
I'd prefer to revert it. Introducing yet another property (there are already far too many) just adds a new way for people to break things.

**Comment by robertmathews** [comment]
> I'd prefer to revert it. Introducing yet another property (there are already far too many) just adds a new way for people to break things.

I agree: It seems better to keep it working as it was without introducing new possibilities for things to go wrong. Before this change, in cases where it didn't previously work, it never worked, and people should just fix their setup as always. In cases where it did previously work even if set up non-optimally, there's no advantage to making it fail by default or giving people an option to make that happen.

I'll also say that although I see the appeal of trying to make SMTP and sendmail work the same way, one could argue that they're not actually the same. With SMTP, there's no default envelope sender that might be available from the environment, so you need to generate one somehow, and taking the address from the From the header is a reasonable thing to do.

But with sendmail, there are other possibilities: if PHPMailer doesn't specify an envelope sender, sendmail might be able to (and in many cases does) find a reasonable value from the php.ini sendmail_from variable, or from the Unix username@hostname. So I think the previous behavior treating them differently actually made sense.

Reverting it seems to be the best bet, and I will submit a PR soon.

**Comment by Synchro** [comment]
This is fixed in the 6.4.0 release.

**Comment by audrasjb** [comment]
Many thanks all!
